sui-execution/latest/sui-adapter/src/static_programmable_transactions/typing/translate.rs

```diff
@@ -0,0 +1,635 @@
+    command_arg_idx: usize,
+    location: T::Location,
+    expected_ty: Type,
+) -> Result<T::Argument, ExecutionError> {
+    let arg__ = argument_(env, context, command_arg_idx, location, &expected_ty)
+        .map_err(|e| e.into_execution_error(command_arg_idx))?;
+    let arg_ = (arg__, expected_ty);
+    Ok(sp(command_arg_idx as u16, arg_))
+}
+
+fn argument_(
+    env: &Env,
+    context: &mut Context,
+    command_arg_idx: usize,
+    location: T::Location,
+    expected_ty: &Type,
+) -> Result<T::Argument__, EitherError> {
+    let command_and_arg_idx = (context.current_command, command_arg_idx as u16);
+    let actual_ty = context.location_type(env, location)?;
+    Ok(match (actual_ty, expected_ty) {
+        // Reference location types
+        (LocationType::Fixed(Type::Reference(a_is_mut, a)), Type::Reference(b_is_mut, b))
+            if !b_is_mut || a_is_mut =>
+        {
+            debug_assert!(!a_is_mut || *b_is_mut);
+            debug_assert!(expected_ty.abilities().has_copy());
+            check_type(command_and_arg_idx, LocationType::Fixed((*a).clone()), b)?;
+            T::Argument__::new_copy(location)
+        }
+        (LocationType::Fixed(Type::Reference(_, a)), b) => {
+            check_type(command_and_arg_idx, LocationType::Fixed((*a).clone()), b)?;
+            if !b.abilities().has_copy() {
+                // TODO this should be a different error for missing copy
+                return Err(CommandArgumentError::TypeMismatch.into());
+            }
+            T::Argument__::Read(T::Usage::new_copy(location))
+        }
+
+        // Non reference location types
+        (actual_ty, Type::Reference(is_mut, inner)) => {
+            check_type_impl(
+                command_and_arg_idx,
+                /* fix */ *is_mut,
+                actual_ty,
+                inner,
+            )?;
+            T::Argument__::Borrow(/* mut */ false, location)
+        }
+        (actual_ty, _) => {
+            check_type(command_and_arg_idx, actual_ty, expected_ty)?;
+            T::Argument__::Use(if expected_ty.abilities().has_copy() {
+                T::Usage::new_copy(location)
+            } else {
+                T::Usage::new_move(location)
+            })
+        }
+    })
+}
+
+fn check_type(
+    command_and_arg_idx: (u16, u16),
+    actual_ty: LocationType,
+    expected_ty: &Type,
+) -> Result<(), CommandArgumentError> {
+    check_type_impl(
+        command_and_arg_idx,
+        /* fix */ false,
+        actual_ty,
+        expected_ty,
+    )
+}
+
+fn check_type_impl(
+    command_and_arg_idx: (u16, u16),
+    fix: bool,
+    actual_ty: LocationType,
+    expected_ty: &Type,
+) -> Result<(), CommandArgumentError> {
+    match actual_ty {
+        LocationType::Bytes(ty, types) => {
+            types
+                .entry(expected_ty.clone())
+                .or_insert(command_and_arg_idx);
+            if fix {
+                *ty = InputType::Fixed(expected_ty.clone());
```

Comment:
What if you have a `LocationType::Bytes(InputType::Fixed(...), ...)` coming in here? I think that should be a type mismatch error as well unless that is checked elsewhere? (but I can't seem to find it).

Reply:
Fantastic! good catch